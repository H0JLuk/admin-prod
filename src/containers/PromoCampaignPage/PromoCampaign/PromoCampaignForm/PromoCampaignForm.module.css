.wrapper {
    width: 100%;
    height: 100%;
}

.promoCreateContainer {
    display: flex;
}

.containerControl {
    position: sticky;
    top: 63px;
    z-index: 1;
    background-color: #ffffff;
    width: 100%;
    padding: 15px;
    margin-bottom: 15px;
    border-bottom: 2px solid #e5e8eb;
}

.containerControl.noMargin {
    margin-bottom: 0;
}

.headerTitle {
    font-size: 20px;
    line-height: 1.4;
    letter-spacing: -0.38px;
    font-weight: 600;
}

.headerSteps {
    font-size: 16px;
    color: #868d98;
    margin-bottom: 10px;
}

.btnMR {
    margin-right: 15px;
}

.error {
    margin-top: 10px;
    color: tomato;
}

.loadingWrapper {
    width: 100%;
    position: absolute;
    height: 100%;
    z-index: 1;
    background-color: rgba(255, 255, 255, 0.5);
}

.loading {
    display: flex;
    align-items: center;
    position: absolute;
    left: 50%;
    top: 0;
    height: 100%;
}

.containerStep {
    padding: 0 15px;
}

.containerStep :global(.ant-form-item-label) {
    text-align: left;
    overflow: visible;
}