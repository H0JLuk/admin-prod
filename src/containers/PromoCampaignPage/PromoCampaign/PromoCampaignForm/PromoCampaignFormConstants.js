export const STEP = 'Шаг';
export const CANCEL = 'Отменить';
export const NEXT = 'Далее';
export const COMPLETE = 'Готово';
export const IMAGE = 'image';
export const BANNER_REQUEST = 'bannerRequest';
export const APPLICATION_JSON_TYPE = 'application/json';
export const allStep = 3;
export const SAVE = 'Сохранить';

export const modes = {
    create: 'create',
    edit: 'edit',
};

export const steps = {
    main_info: 1,
    textAndBanners: 2,
    visibility: 3
};

export const pageTypes = {
    present: 'PRESENT',
    normal: 'NORMAL',
};

export const modsTitle = {
    create: 'Новая промо-кампания',
    edit: 'Редактирование промо-кампании',
};

export const promoCampaignTextTypes = {
    RULES: 'RULES',
    HEADER: 'HEADER',
    DESCRIPTION: 'DESCRIPTION',
};

export const promoCampaignBannerTypes = {
    CARD: 'CARD',
    LOGO_MAIN: 'LOGO_MAIN',
    LOGO_ICON: 'LOGO_ICON',
    SCREEN: 'SCREEN',
    LOGO_SECONDARY: 'LOGO_SECONDARY',
};

export const bannersFields = 'banners';