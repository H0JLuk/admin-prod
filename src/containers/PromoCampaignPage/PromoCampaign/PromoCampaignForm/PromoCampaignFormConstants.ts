export const STEP = 'Шаг';
export const CANCEL = 'Отменить';
export const NEXT = 'Далее';
export const COMPLETE = 'Готово';
export const IMAGE = 'image';
export const BANNER_REQUEST = 'bannerRequest';
export const allStep = 3;
export const SAVE = 'Сохранить';

export const modes = {
    create: 'create',
    edit: 'edit',
};

export const steps = {
    main_info: 1,
    textAndBanners: 2,
    visibility: 3
};

export const pageTypes = {
    present: 'PRESENT',
    normal: 'NORMAL',
};

export const modsTitle = {
    create: 'Новая промо-кампания',
    edit: 'Редактирование промо-кампании',
};

export const bannersFields = 'banners';
