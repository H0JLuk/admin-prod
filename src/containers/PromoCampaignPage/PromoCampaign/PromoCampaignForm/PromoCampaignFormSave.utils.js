export function filterBanner (promoCampaign, bannersIdList, textsList) {
    const textsIDs = textsList.map((text) => text.id);
    const texts = promoCampaign.texts.filter((text) => !textsIDs.includes(text.id));
    const banners = promoCampaign.banners.filter((banner) => !bannersIdList.includes(banner.id));

    return { ...promoCampaign, banners, texts };
}

export function deleteText(promoCampaign, id) {
    const texts = promoCampaign.texts.filter((text) => text.id !== id);
    return { ...promoCampaign, texts };
}
