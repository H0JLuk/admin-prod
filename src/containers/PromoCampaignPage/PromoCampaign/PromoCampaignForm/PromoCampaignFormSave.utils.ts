import { PromoCampaignDto } from '@types';

export function filterBanner(promoCampaign: PromoCampaignDto, bannersIdList: number[], textsList: PromoCampaignDto['texts']) {
    const textsIDs = textsList.map((text) => text.id);
    const texts = promoCampaign.texts.filter((text) => !textsIDs.includes(text.id));
    const banners = promoCampaign.banners.filter((banner) => !bannersIdList.includes(banner.id));

    return { ...promoCampaign, banners, texts };
}

export function deleteText(promoCampaign: PromoCampaignDto, id: number) {
    const texts = promoCampaign.texts.filter((text) => text.id !== id);
    return { ...promoCampaign, texts };
}
