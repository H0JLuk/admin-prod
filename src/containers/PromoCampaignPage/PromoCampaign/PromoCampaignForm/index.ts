import PromoCampaignForm from './PromoCampaignForm';

export type { PromoCampaignFormProps } from './PromoCampaignForm';

export default PromoCampaignForm;
