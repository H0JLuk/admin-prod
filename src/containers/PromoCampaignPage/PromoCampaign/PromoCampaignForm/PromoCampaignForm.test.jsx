import React from 'react';
import { shallow, mount } from 'enzyme';
import PromoCampaignForm from './PromoCampaignForm';
import callConfirmModalForPromoCodeTypeChanging from './PromoCampaignSteps/ConfirmModalForPromoCodeTypeChanging/ConfirmModalForPromoCodeTypeChanging';
import { editPromoCampaign, newPromoCampaign, copyPromoCampaign } from '../../../../api/services/promoCampaignService';
import { getUnissuedPromoCodeStatistics } from '../../../../api/services/promoCodeService';
import { filterBanner, deleteText } from './PromoCampaignFormSave.utils';
import {
    createImgBanners,
    createTexts,
    createVisibilities,
    EditImgBanners,
    editTextBanners,
    getDataForSend,
    normalizePromoCampaignData,
    getPromoCampaignValue,
    checkPromoCodes,
} from './PromoCampaignFormUtils';
import {
    promoCampaignTestData,
    visibilitySettingLocation,
} from '../../../../../__tests__/constants';
import { sleep } from '../../../../setupTests';


let mockEmptyStateLocation = false;
let mockCopyVisibilitySettings = false;
const mockState = promoCampaignTestData;
const mockHistoryPush = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
        push: mockHistoryPush,
    }),
    useLocation: () => ({
        state: !mockEmptyStateLocation ? {
            promoCampaign: mockState,
            copyVisibilitySettings: mockCopyVisibilitySettings,
        } : undefined,
    }),
}));

jest.mock(
    './PromoCampaignSteps/ConfirmModalForPromoCodeTypeChanging/ConfirmModalForPromoCodeTypeChanging',
    () => jest.fn()
);

jest.mock('../../../../api/services/promoCampaignService', () => ({
    editPromoCampaign: jest.fn(),
    newPromoCampaign: jest.fn(),
    copyPromoCampaign: jest.fn(),
}));

jest.mock('../../../../api/services/promoCodeService', () => ({
    getUnissuedPromoCodeStatistics: jest.fn(),
}));

jest.mock('./PromoCampaignFormSave.utils', () => ({
    filterBanner: jest.fn(),
    deleteText: jest.fn(),
}));

jest.mock('./PromoCampaignFormUtils', () => ({
    ...jest.requireActual('./PromoCampaignFormUtils'),
    checkPromoCodes: jest.fn(),
    createImgBanners: jest.fn(),
    createTexts: jest.fn(),
    createVisibilities: jest.fn(),
    EditImgBanners: jest.fn(),
    editTextBanners: jest.fn(),
    getDataForSend: jest.fn(),
    getPromoCampaignForCopy: jest.fn(),
    normalizeFirstStepValue: jest.fn(),
    normalizePromoCampaignData: jest.fn(),
    getPromoCampaignValue: jest.fn(),
}));

jest.mock('./PromoCampaignSteps/StepInfo/StepInfo', () =>
    /* Сделано именно так, потому что нам нужно чтобы в редеринге enzyme отображалось имя компонента */
    function StepInfo() {
        return <div>StepInfo</div>;
    }
);

const testProps = {
    matchPath: '/test/path',
};

describe('<PromoCampaignForm.test.jsx', () => {
    const wrapper = shallow(<PromoCampaignForm { ...testProps } />);

    beforeEach(() => {
        wrapper.find('PromoCampaignSideBar').simulate('click', 1);
    });

    it('should render', () => {
        expect(wrapper.find('.headerTitle').text()).toBe('Новая промо-кампания');
        expect(wrapper.debug()).toMatchSnapshot();
    });

    it('should render <StepInfo />', () => {
        expect(wrapper.find('StepInfo')).toHaveLength(1);
        expect(wrapper.find('StepTextAndImage')).toHaveLength(0);
        expect(wrapper.find('StepVisibility')).toHaveLength(0);
        expect(wrapper.find('.headerSteps').text()).toBe('Шаг 1 из 3');
        expect(wrapper.find('Button')).toHaveLength(3);
    });

    it('should render <StepTextAndImage />', () => {
        wrapper.find('PromoCampaignSideBar').simulate('click', 2);
        expect(wrapper.find('StepInfo')).toHaveLength(0);
        expect(wrapper.find('StepTextAndImage')).toHaveLength(1);
        expect(wrapper.find('StepVisibility')).toHaveLength(0);
        expect(wrapper.find('.headerSteps').text()).toBe('Шаг 2 из 3');
        expect(wrapper.find('Button')).toHaveLength(3);
    });

    it('should render <StepVisibility />', () => {
        wrapper.find('PromoCampaignSideBar').simulate('click', 3);
        expect(wrapper.find('StepInfo')).toHaveLength(0);
        expect(wrapper.find('StepTextAndImage')).toHaveLength(0);
        expect(wrapper.find('StepVisibility')).toHaveLength(1);
        expect(wrapper.find('.headerSteps').text()).toBe('Шаг 3 из 3');
        expect(wrapper.find('Button')).toHaveLength(2);
    });

    it('should render null inside <Form />', () => {
        wrapper.find('PromoCampaignSideBar').simulate('click', 0);
        const formWrapper = wrapper.find('ForwardRef(InternalForm)');
        expect(formWrapper.children()).toHaveLength(0);
    });

    it('should return to previous step', () => {
        wrapper.find('PromoCampaignSideBar').simulate('click', 2);
        expect(wrapper.find('PromoCampaignSideBar').prop('active')).toBe(2);

        wrapper.find('Header').simulate('clickFunc');
        expect(wrapper.find('PromoCampaignSideBar').prop('active')).toBe(1);
    });

    it('should call history.push() when click cancel button', () => {
        wrapper.find('Button').first().simulate('click');
        expect(mockHistoryPush).toBeCalledWith(testProps.matchPath);
    });

    it('should change validStep', async () => {
        console.error = jest.fn();
        expect(wrapper.find('PromoCampaignSideBar').prop('validStep')).toBe(1);
        expect(wrapper.find('PromoCampaignSideBar').prop('active')).toBe(1);
        wrapper.find('Button').at(1).simulate('click');
        await sleep();

        expect(wrapper.find('PromoCampaignSideBar').prop('validStep')).toBe(2);
        expect(wrapper.find('PromoCampaignSideBar').prop('active')).toBe(2);
    });

    it('should change validStep and step', async () => {
        wrapper.find('PromoCampaignSideBar').simulate('click', 2);
        expect(wrapper.find('PromoCampaignSideBar').prop('validStep')).toBe(2);
        expect(wrapper.find('PromoCampaignSideBar').prop('active')).toBe(2);

        wrapper.find('Button').at(1).simulate('click');
        await sleep();

        expect(wrapper.find('PromoCampaignSideBar').prop('validStep')).toBe(3);
        expect(wrapper.find('PromoCampaignSideBar').prop('active')).toBe(3);
    });

    it('check visibilitySettings empty error', () => {
        wrapper.find('PromoCampaignSideBar').simulate('click', 3);
        expect(wrapper.find('StepVisibility').prop('visibilitySettings')[0].errors).toEqual({});

        wrapper.find('Button').at(1).simulate('click');
        expect(wrapper.find('StepVisibility').prop('visibilitySettings')[0].errors).toEqual({ location: 'Укажите локацию' });
    });

    it('check visibilitySettings same settings error', () => {
        wrapper.find('PromoCampaignSideBar').simulate('click', 3);
        const StepVisibility = wrapper.find('StepVisibility');
        expect(StepVisibility.prop('visibilitySettings')[0].errors).toEqual({ location: 'Укажите локацию' });

        StepVisibility.simulate('changeState', visibilitySettingLocation, 0, 'location', {});

        expect(
            wrapper.find('StepVisibility').prop('visibilitySettings')[0].location
        ).toEqual(visibilitySettingLocation);

        wrapper.find('StepVisibility').simulate(
            'changeState',
            { location: null, salePoint: null, visible: false, errors: {}, id: Date.now() },
            1
        );

        expect(wrapper.find('StepVisibility').prop('visibilitySettings')).toHaveLength(2);

        wrapper.find('StepVisibility').simulate('changeState', visibilitySettingLocation, 1, 'location', {});

        wrapper.find('Button').at(1).simulate('click');
        const [firstSetting, secondSetting] = wrapper.find('StepVisibility').prop('visibilitySettings');
        expect(firstSetting.errors).toEqual({
            server: 'Нельзя добавить одинаковые настройки видимости с локацией \'Ростов\' ',
        });
        expect(secondSetting.errors).toEqual({
            server: 'Нельзя добавить одинаковые настройки видимости с локацией \'Ростов\' ',
        });
    });

    it('should delete second visibilitySetting', () => {
        wrapper.find('PromoCampaignSideBar').simulate('click', 3);

        wrapper.find('StepVisibility').simulate('deleteState', undefined);
        expect(wrapper.find('StepVisibility').prop('visibilitySettings')).toHaveLength(2);

        wrapper.find('StepVisibility').simulate('deleteState', 1, true);
        const visSettings = wrapper.find('StepVisibility').prop('visibilitySettings');
        expect(visSettings).toHaveLength(1);
        expect(visSettings[0].errors).toEqual({});
    });

    it('should call `newPromoCampaign` function and history.push', async () => {
        newPromoCampaign.mockReturnValue({ id: 99 });
        createTexts.mockReturnValue([]);
        createVisibilities.mockReturnValue([]);
        createImgBanners.mockResolvedValue();
        wrapper.find('PromoCampaignSideBar').simulate('click', 3);
        wrapper.find('Button').at(1).simulate('click');
        await sleep();

        expect(newPromoCampaign).toBeCalledTimes(1);
        expect(mockHistoryPush).toBeCalledTimes(1);
        expect(mockHistoryPush).toBeCalledWith(testProps.matchPath);
    });

    it('save promoCampaign should rejected and render error', async () => {
        newPromoCampaign.mockRejectedValue(new Error('test Error when create'));
        wrapper.find('PromoCampaignSideBar').simulate('click', 3);
        wrapper.find('Button').at(1).simulate('click');
        await sleep();

        expect(newPromoCampaign).toBeCalledTimes(1);
        expect(mockHistoryPush).toBeCalledTimes(0);
        expect(wrapper.find('.error').text()).toBe('test Error when create');
    });

    it('should change valid step when form value was changed', async () => {
        wrapper.find('Button').at(1).simulate('click');
        await sleep();
        wrapper.find('PromoCampaignSideBar').simulate('click', 1);

        expect(wrapper.find('PromoCampaignSideBar').prop('active')).toBe(1);
        expect(wrapper.find('PromoCampaignSideBar').prop('validStep')).toBe(2);

        wrapper.find('ForwardRef(InternalForm)').simulate('valuesChange', {
            banners: {
                ICON_MAIN: 'test_value',
            },
        });
        expect(wrapper.find('PromoCampaignSideBar').prop('active')).toBe(1);
        expect(wrapper.find('PromoCampaignSideBar').prop('validStep')).toBe(1);
    });

    it('should reject `newPromoCampaign` function when click on Save btn', async () => {
        newPromoCampaign.mockRejectedValue(new Error('new test Error when create'));
        getDataForSend.mockReturnValue({ test: 'test data for send' });
        wrapper.find('ForwardRef(InternalForm)').simulate('valuesChange', {
            test: 'test_change',
        });
        wrapper.find('Button').last().simulate('click');
        await sleep();

        expect(newPromoCampaign).toBeCalledTimes(1);
        expect(wrapper.find('.error').text()).toBe('new test Error when create');
    });

    it('should call `newPromoCampaign` function when click on Save btn', async () => {
        newPromoCampaign.mockReturnValue({ id: 99 });
        getDataForSend.mockReturnValue({ test: 'test data for send' });
        wrapper.find('Button').last().simulate('click');
        await sleep();

        expect(newPromoCampaign).toBeCalledTimes(1);
    });

    it('should reject call `editPromoCampaign` function', async () => {
        editPromoCampaign.mockRejectedValue(new Error('test reject edit'));
        wrapper.find('ForwardRef(InternalForm)').simulate('valuesChange', {
            test: 'test_change',
        });
        wrapper.find('Button').last().simulate('click');

        await sleep();

        expect(editPromoCampaign).toBeCalledTimes(1);
        expect(wrapper.find('.error').text()).toBe('test reject edit');
    });

    it('should call `editPromoCampaign` function', async () => {
        getDataForSend.mockReturnValue({ test: 'test data for send' });
        wrapper.find('Button').last().simulate('click');

        await sleep();
        expect(editPromoCampaign).toBeCalledTimes(1);
    });

    it('should call `callConfirmModalForPromoCodeTypeChanging` function', async () => {
        checkPromoCodes.mockReturnValue('newCodeType');
        getDataForSend.mockReturnValue({ test: 'test data for send', promoCodeType: 'testType' });
        callConfirmModalForPromoCodeTypeChanging.mockReturnValue(true);
        wrapper.find('ForwardRef(InternalForm)').simulate('valuesChange', {
            test: 'test_change',
        });
        wrapper.find('Button').last().simulate('click');

        await sleep();

        getPromoCampaignValue.mockReturnValue({ id: 99 });
        expect(callConfirmModalForPromoCodeTypeChanging).toBeCalledTimes(1);
        const [callback] = callConfirmModalForPromoCodeTypeChanging.mock.calls[0];
        await callback();

        expect(editPromoCampaign).toBeCalledTimes(1);
        expect(getUnissuedPromoCodeStatistics).toBeCalledTimes(1);
        expect(getUnissuedPromoCodeStatistics).toBeCalledWith(99, 'testType');
    });

    it('should call `editTextBanners` and `EditImgBanners`', async () => {
        getDataForSend.mockReturnValue({ test: 'test data for send' });
        editTextBanners.mockReturnValue([]);
        filterBanner.mockReturnValue({
            banners: [],
            texts: [],
        });
        EditImgBanners.mockResolvedValue({
            banners: [],
            deletedBannersId: [],
        });
        wrapper.find('PromoCampaignSideBar').simulate('click', 2);
        wrapper.find('ForwardRef(InternalForm)').simulate('valuesChange', {
            test: 'test_change',
        });
        wrapper.find('Button').last().simulate('click');

        await sleep();

        expect(editTextBanners).toBeCalledTimes(1);
        expect(EditImgBanners).toBeCalledTimes(1);
        expect(filterBanner).toBeCalledTimes(1);

        const callback = editTextBanners.mock.calls[0][3];
        deleteText.mockReturnValue({ texts: [] });
        callback(1, 'HEADER');
        expect(deleteText).toBeCalledTimes(1);
    });

    it('should call `editTextBanners`, `EditImgBanners` and newPromoCampaign functions', async () => {
        const wrapper = shallow(<PromoCampaignForm { ...testProps } />);
        newPromoCampaign.mockReturnValue({ id: 98 });
        getDataForSend.mockReturnValue({ test: 'test data for send' });
        editTextBanners.mockReturnValue([]);
        filterBanner.mockReturnValue({
            banners: [],
            texts: [],
        });
        EditImgBanners.mockResolvedValue({
            banners: [],
            deletedBannersId: [],
        });
        wrapper.find('PromoCampaignSideBar').simulate('click', 2);
        wrapper.find('ForwardRef(InternalForm)').simulate('valuesChange', {
            test: 'test_change',
        });
        wrapper.find('Button').last().simulate('click');

        await sleep();

        expect(editTextBanners).toBeCalledTimes(1);
        expect(EditImgBanners).toBeCalledTimes(1);
        expect(filterBanner).toBeCalledTimes(1);
        expect(newPromoCampaign).toBeCalledTimes(1);
    });

    it('should call `callConfirmModalForPromoCodeTypeChanging` function when edit', async () => {
        checkPromoCodes.mockReturnValue('newTestType');
        getPromoCampaignValue.mockReturnValue({ id: 92 });
        getDataForSend.mockReturnValue({ banners: [], texts: [] });
        editTextBanners.mockReturnValue([]);
        createVisibilities.mockReturnValue([]);
        EditImgBanners.mockResolvedValue({
            banners: [],
            deletedBannersId: [],
        });
        wrapper.find('PromoCampaignSideBar').simulate('click', 3);
        wrapper.find('Button').last().simulate('click');
        await sleep();

        expect(callConfirmModalForPromoCodeTypeChanging).toBeCalledTimes(1);
        const [callback] = callConfirmModalForPromoCodeTypeChanging.mock.calls[0];
        await callback();

        expect(getUnissuedPromoCodeStatistics).toBeCalledTimes(1);
        expect(editPromoCampaign).toBeCalledTimes(1);
        expect(editTextBanners).toBeCalledTimes(1);
        expect(EditImgBanners).toBeCalledTimes(1);
        expect(createVisibilities).toBeCalledTimes(1);
        expect(mockHistoryPush).toBeCalledTimes(1);
    });

    it('should call edit functions. First time - reject, second - resolve', async () => {
        checkPromoCodes.mockReturnValue(null);
        getPromoCampaignValue.mockReturnValue({ id: 92 });
        getDataForSend.mockReturnValue({ banners: [], texts: [] });
        editTextBanners.mockReturnValue([]);
        createVisibilities.mockReturnValue([]);
        EditImgBanners.mockResolvedValue({
            banners: [],
            deletedBannersId: [],
        });
        editPromoCampaign
            .mockRejectedValueOnce(new Error('error while edit'))
            .mockResolvedValueOnce('success edit');

        wrapper.find('PromoCampaignSideBar').simulate('click', 3);
        wrapper.find('Button').last().simulate('click');
        await sleep();

        // expect calls of functions when reject
        expect(callConfirmModalForPromoCodeTypeChanging).toBeCalledTimes(0);
        expect(getUnissuedPromoCodeStatistics).toBeCalledTimes(0);
        expect(editPromoCampaign).toBeCalledTimes(1);
        expect(editTextBanners).toBeCalledTimes(0);
        expect(EditImgBanners).toBeCalledTimes(0);
        expect(createVisibilities).toBeCalledTimes(0);
        expect(mockHistoryPush).toBeCalledTimes(0);
        expect(wrapper.find('.error').text()).toBe('error while edit');

        wrapper.find('PromoCampaignSideBar').simulate('click', 3);
        wrapper.find('Button').last().simulate('click');
        await sleep();

        // expect calls of functions when resolve
        expect(callConfirmModalForPromoCodeTypeChanging).toBeCalledTimes(0);
        expect(getUnissuedPromoCodeStatistics).toBeCalledTimes(0);
        expect(editPromoCampaign).toBeCalledTimes(2);
        expect(editTextBanners).toBeCalledTimes(1);
        expect(EditImgBanners).toBeCalledTimes(1);
        expect(createVisibilities).toBeCalledTimes(1);
        expect(mockHistoryPush).toBeCalledTimes(1);
    });

    it('should don`t call functions with empty visibilitySettings', async () => {
        checkPromoCodes.mockReturnValue(null);
        wrapper.find('PromoCampaignSideBar').simulate('click', 3);
        wrapper.find('StepVisibility').simulate('changeState', null, 0, 'location');
        wrapper.find('Button').last().simulate('click');
        await sleep();

        expect(callConfirmModalForPromoCodeTypeChanging).toBeCalledTimes(0);
        expect(getUnissuedPromoCodeStatistics).toBeCalledTimes(0);
        expect(editPromoCampaign).toBeCalledTimes(0);
        expect(editTextBanners).toBeCalledTimes(0);
        expect(EditImgBanners).toBeCalledTimes(0);
        expect(createVisibilities).toBeCalledTimes(0);
        expect(mockHistoryPush).toBeCalledTimes(0);
        expect(
            wrapper.find('StepVisibility').prop('visibilitySettings')[0].errors
        ).toEqual({ location: 'Укажите локацию' });
    });

    it('should clear texts and banners in state', () => {
        wrapper.find('StepInfo').prop('changeTypePromo')();
        wrapper.find('PromoCampaignSideBar').simulate('click', 2);
        expect(wrapper.find('StepTextAndImage').prop('texts')).toEqual({});
        expect(wrapper.find('StepTextAndImage').prop('banners')).toEqual({});
    });

    describe('<PromoCampaignForm /> tests with `isCopy = true` prop', () => {
        const wrapper = shallow(
            <PromoCampaignForm
                { ...testProps }
                mode="edit"
                isCopy
            />
        );

        beforeEach(() => {
            wrapper.find('PromoCampaignSideBar').simulate('click', 1);
            getDataForSend.mockReturnValue({ banners: [], texts: [] });
            editTextBanners.mockReturnValue([]);
            createVisibilities.mockReturnValue([]);
            EditImgBanners.mockResolvedValue({
                banners: [],
                deletedBannersId: [],
            });
        });

        it('check function `handleNextStep` and `copyVisibilitySettings = true`', async () => {
            mockCopyVisibilitySettings = true;
            wrapper.update();
            copyPromoCampaign.mockReturnValue(promoCampaignTestData);
            getPromoCampaignValue.mockReturnValue({ id: 11 });

            wrapper.find('PromoCampaignSideBar').simulate('click', 2);
            wrapper.find('Button').at(1).simulate('click');
            await sleep();

            expect(copyPromoCampaign).toBeCalledTimes(1);
            expect(editPromoCampaign).toBeCalledTimes(0);
            expect(editTextBanners).toBeCalledTimes(1);
            expect(EditImgBanners).toBeCalledTimes(1);
            expect(createVisibilities).toBeCalledTimes(0);
            expect(mockHistoryPush).toBeCalledTimes(1);

            mockCopyVisibilitySettings = false;
            wrapper.update();
        });

        it('check function `handleNextStep` and `copyVisibilitySettings = false`. should not call without visibilitySettings', async () => {
            wrapper.find('PromoCampaignSideBar').simulate('click', 3);
            expect(
                wrapper.find('StepVisibility').prop('visibilitySettings')[0].errors
            ).toEqual({});
            wrapper.find('Button').at(1).simulate('click');
            await sleep();

            expect(copyPromoCampaign).toBeCalledTimes(0);
            expect(editPromoCampaign).toBeCalledTimes(0);
            expect(editTextBanners).toBeCalledTimes(0);
            expect(EditImgBanners).toBeCalledTimes(0);
            expect(createVisibilities).toBeCalledTimes(0);
            expect(mockHistoryPush).toBeCalledTimes(0);
            expect(
                wrapper.find('StepVisibility').prop('visibilitySettings')[0].errors
            ).toEqual({ location: 'Укажите локацию' });
        });

        it('check function `handleNextStep` and `copyVisibilitySettings = false`', async () => {
            copyPromoCampaign.mockReturnValue(promoCampaignTestData);
            getPromoCampaignValue.mockReturnValue({ id: 11 });

            wrapper.find('PromoCampaignSideBar').simulate('click', 3);
            wrapper.find('StepVisibility').simulate('changeState', visibilitySettingLocation, 0, 'location', {});
            wrapper.find('Button').at(1).simulate('click');
            await sleep();

            expect(copyPromoCampaign).toBeCalledTimes(1);
            expect(editPromoCampaign).toBeCalledTimes(0);
            expect(editTextBanners).toBeCalledTimes(1);
            expect(EditImgBanners).toBeCalledTimes(1);
            expect(createVisibilities).toBeCalledTimes(1);
            expect(mockHistoryPush).toBeCalledTimes(1);
        });

        it('should call `copyPromoCampaign` when click saveBtn', async () => {
            copyPromoCampaign.mockReturnValue(promoCampaignTestData);
            getPromoCampaignValue.mockReturnValue({ id: 11 });
            normalizePromoCampaignData.mockReturnValue({
                ...promoCampaignTestData,
                banners: {},
                texts: {},
            });

            wrapper.find('Button').last().simulate('click');
            await sleep();

            expect(copyPromoCampaign).toBeCalledTimes(1);
            expect(editPromoCampaign).toBeCalledTimes(0);

            wrapper.find('ForwardRef(InternalForm)').simulate('valuesChange', {
                test: 'test_change',
            });
            wrapper.find('Button').last().simulate('click');
            await sleep();

            expect(editPromoCampaign).toBeCalledTimes(1);
            expect(copyPromoCampaign).toBeCalledTimes(1);
        });

        it('should call `editPromoCampaign` after success copy and next click Done Btn', async () => {
            getPromoCampaignValue.mockReturnValue(promoCampaignTestData);
            checkPromoCodes.mockReturnValue(null);
            createVisibilities.mockReturnValue([]);
            wrapper.find('PromoCampaignSideBar').simulate('click', 3);
            wrapper.find('Button').last().simulate('click');
            await sleep();

            expect(copyPromoCampaign).toBeCalledTimes(0);
            expect(editPromoCampaign).toBeCalledTimes(1);
        });

        it('should call `editPromoCampaign` after success copy and next click Done Btn on second step', async () => {
            getPromoCampaignValue.mockReturnValue(promoCampaignTestData);
            checkPromoCodes.mockReturnValue(null);
            createVisibilities.mockReturnValue([]);
            filterBanner.mockReturnValue({
                banners: [],
                texts: [],
            });
            wrapper.find('PromoCampaignSideBar').simulate('click', 2);
            wrapper.find('Button').last().simulate('click');
            await sleep();

            expect(copyPromoCampaign).toBeCalledTimes(0);
            expect(editPromoCampaign).toBeCalledTimes(1);
            expect(editTextBanners).toBeCalledTimes(1);
            expect(EditImgBanners).toBeCalledTimes(1);
        });
    });

    describe('<PromoCampaignForm /> tests useEffect', () => {

        it('should call history.push', async () => {
            mockEmptyStateLocation = true;
            mount(<PromoCampaignForm { ...testProps } mode="edit" />);

            expect(mockHistoryPush).toBeCalledTimes(1);
            expect(mockHistoryPush).toBeCalledWith(testProps.matchPath);
            mockEmptyStateLocation = false;
        });

        it('should don`t render loading block and validStep = 1', () => {
            const wrapper = mount(<PromoCampaignForm { ...testProps } />);
            expect(wrapper.find('.loadingWrapper')).toHaveLength(0);
            expect(wrapper.find('PromoCampaignSideBar').prop('validStep')).toBe(1);
        });

        it('should don`t render loading block and validStep = 3', () => {
            normalizePromoCampaignData.mockReturnValue(promoCampaignTestData);
            const wrapper = mount(<PromoCampaignForm { ...testProps } mode="edit" />);
            expect(wrapper.find('.loadingWrapper')).toHaveLength(0);
            expect(wrapper.find('PromoCampaignSideBar').prop('validStep')).toBe(3);
        });

    });
});
