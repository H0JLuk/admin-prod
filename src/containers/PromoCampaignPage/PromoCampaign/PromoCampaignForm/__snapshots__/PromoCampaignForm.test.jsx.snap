// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PromoCampaignForm /> tests should render 1`] = `
"<div className=\\"promoCreateContainer\\">
  <PromoCampaignSideBar active={1} validStep={1} onClick={[Function: bound _dispatchAction]} newPromoCampaign={true} hideLastStep={false} />
  <div className=\\"wrapper\\">
    <Header onClickFunc={[undefined]} />
    <div className=\\"containerControl\\">
      <p className=\\"headerTitle\\">
        Новая промо-кампания
      </p>
      <p className=\\"headerSteps\\">
        Шаг
         
        1
         из 
        3
      </p>
      <Button className=\\"btnMR\\" onClick={[Function: handleCancel]}>
        Отменить
      </Button>
      <Button type=\\"primary\\" className=\\"btnMR\\" onClick={[Function: nextStep]}>
        Далее
      </Button>
      <Button type=\\"primary\\" onClick={[Function: handleSaveInfo]} disabled={true}>
        Сохранить
      </Button>
    </div>
    <ForwardRef(InternalForm) className=\\"containerStep\\" form={{...}} layout=\\"vertical\\" onValuesChange={[Function: isChanged]}>
      <StepInfo isCopy={[undefined]} state={{...}} changeTypePromo={[Function (anonymous)]} oldName=\\"Промо-кампания Сбермаркет для витрины ВСП\\" oldExternalId={{...}} copyPromoCampaignId={{...}} mode=\\"create\\" />
    </ForwardRef(InternalForm)>
  </div>
  <div className=\\"loadingWrapper\\">
    <div className=\\"loading\\">
      <ForwardRef(SvgLoadingSpinner) />
    </div>
  </div>
</div>"
`;
